File: ios/Shared/WatchPayload.swift
import Foundation

/// What the phone hands the watch.
///
/// Not the whole document — the watch can't edit anything, so it only gets what it
/// draws, with the dates already formatted. Compiled into both targets so the two
/// ends can't drift apart.
///
/// Two shapes ride together: `sections` is the original reminders-by-group list, and
/// `days` is the week window (today first, seven days) the watch's Today / Reminders /
/// Events / All views and the complication draw from. Both ends decode tolerantly, so
/// a phone and a watch on different versions keep working.
struct WatchList: Codable, Hashable {
    var folder = ""
    var sections: [WatchSection] = []
    var days: [WatchDay] = []

    init(folder: String = "", sections: [WatchSection] = [], days: [WatchDay] = []) {
        self.folder = folder
        self.sections = sections
        self.days = days
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        folder   = try c.decodeIfPresent(String.self, forKey: .folder) ?? ""
        sections = try c.decodeIfPresent([WatchSection].self, forKey: .sections) ?? []
        days     = try c.decodeIfPresent([WatchDay].self, forKey: .days) ?? []
    }
}

struct WatchSection: Codable, Hashable, Identifiable {
    var id: String { name }
    var name: String
    var items: [WatchItem]
}

/// One day of the week window: its items in the day panel's order — events (by time),
/// then reminders (undated-first, then date, then time), then notes.
struct WatchDay: Codable, Hashable, Identifiable {
    var id: String        // "2026-08-03"
    var name: String      // "Today · Aug 3", "Tue · Aug 4"
    var items: [WatchItem]
}

struct WatchItem: Codable, Hashable, Identifiable {
    var id: String
    var text: String
    var due: String         // "today", "2pm", "Aug 3", or ""
    var overdue: Bool
    var kind: String        // "reminder" | "event" | "note"

    init(id: String, text: String, due: String, overdue: Bool, kind: String = "reminder") {
        self.id = id
        self.text = text
        self.due = due
        self.overdue = overdue
        self.kind = kind
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id      = try c.decodeIfPresent(String.self, forKey: .id) ?? UUID().uuidString
        text    = try c.decodeIfPresent(String.self, forKey: .text) ?? ""
        due     = try c.decodeIfPresent(String.self, forKey: .due) ?? ""
        overdue = try c.decodeIfPresent(Bool.self, forKey: .overdue) ?? false
        kind    = try c.decodeIfPresent(String.self, forKey: .kind) ?? "reminder"
    }
}

/// "Today · Aug 3" for today, else "Tue · Aug 4" — the web widget's day headings.
func watchDayName(_ date: Date, today: Date) -> String {
    let f = DateFormatter()
    f.locale = Locale(identifier: "en_US_POSIX")
    f.dateFormat = "MMM d"
    let md = f.string(from: date)
    if Calendar.current.isDate(date, inSameDayAs: today) { return "Today · \(md)" }
    f.dateFormat = "EEE"
    return "\(f.string(from: date)) · \(md)"
}

enum WatchLink {
    /// The single key in the WatchConnectivity application context.
    static let listKey = "list"
    /// The app-group defaults suite the watch app and its complication share.
    static let appGroup = "group.com.seancheren.suite"
    /// The key the decoded list is cached under in those defaults.
    static let cacheKey = "watchList"
}

File: android/core/src/main/kotlin/com/seancheren/suite/core/WatchPayload.kt
package com.seancheren.suite.core

import kotlinx.serialization.Serializable
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * What the phone hands a watch — the cousin of ios/Shared/WatchPayload.swift.
 *
 * Not the whole document — the watch can't edit anything, so it only gets what it
 * draws, with the dates already formatted. Built by Store.watchList() so the two ends
 * can't drift apart. (Wear OS is a later job; this shape is ready for it.)
 */
@Serializable
data class WatchList(
    var folder: String = "",
    var sections: List<WatchSection> = emptyList(),
    var days: List<WatchDay> = emptyList(),
)

/**
 * One day of the week window (today first, seven days): its items in the day panel's
 * order — events (by time), then reminders (undated-first, then date, then time), then
 * notes. The watch's Today / Reminders / Events / All views and the complication draw
 * from these.
 */
@Serializable
data class WatchDay(
    var id: String,       // "2026-08-03"
    var name: String,     // "Today · Aug 3", "Tue · Aug 4"
    var items: List<WatchItem>,
)

@Serializable
data class WatchSection(
    var name: String,
    var items: List<WatchItem>,
) {
    val id: String get() = name
}

@Serializable
data class WatchItem(
    var id: String,
    var text: String,
    var due: String,        // "today", "2pm", "Aug 3", or ""
    var overdue: Boolean,
    var kind: String = "reminder",   // "reminder" | "event" | "note" — defaulted, so an old payload decodes
)

/** "Today · Aug 3" for today, else "Tue · Aug 4" — the web widget's day headings. */
fun watchDayName(date: LocalDate, today: LocalDate): String {
    val md = date.format(DateTimeFormatter.ofPattern("MMM d", Locale.US))
    if (date == today) return "Today · $md"
    return "${date.format(DateTimeFormatter.ofPattern("EEE", Locale.US))} · $md"
}

object WatchLink {
    /** The single key in the phone→watch payload. */
    const val LIST_KEY = "list"
}
